### PR TITLE
Implementation of the example Scala challenge

### PR DESCRIPTION
**Pull request by marcinkrykowski**
Hi guys 👋🏼

The code definitely isn't perfect, and it misses many things 🙄

The reason for that is that I only had 8 hours of free time I could use to create the project. ⏱

I still wanted to show it, so you can have an idea of what I can accomplish in an 8-hour session. 🤯

I wanted to roast myself, so I put some comments here and there. 😂

If you have any questions, please let me know ❗️

PS. I know that git history is messy here. Initial commit contains 90% of implementation.

### CODE REVIEW

File: src/it/scala/com/krykowski/api/TaskApiServerSpec.scala [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      val request = Request[IO](
+        method = Method.GET,
+        uri = Uri.unsafeFromString(
+          s"$urlStart/task"
+        )
+      )
+      val response = client.use(_.expect[String](request)).unsafeRunSync()
+      response should include("[]")
+    }
+  }
```
Comment (marcinkrykowski) [inline src/it/scala/com/krykowski/api/TaskApiServerSpec.scala:64]: No tests for other cases. No `IT` tests for unhappy path.

File: src/main/scala/com/krykowski/model/package.scala [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package com.krykowski
+
+package object model {
+
+  case class Task(id: Option[Long], status: Status = Scheduled, jsonFileUri: String)
```
Comment (marcinkrykowski) [inline src/main/scala/com/krykowski/model/package.scala:5]: The model is not accurate. Does not contain number of processed lines and pace.

File: src/main/scala/com/krykowski/repository/TaskInMemoryRepository.scala [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+  var tasks: List[Task] = List()
+
+  def createTask(task: Task): IO[Long] = {
+    logger.info(s"Creating new task: ${task.toString}")
+    val highestId =
+      if (tasks.isEmpty) 0L
+      else {
+        tasks.maxBy(_.id).id.get
+      }
```
Comment (marcinkrykowski) [inline src/main/scala/com/krykowski/repository/TaskInMemoryRepository.scala:18]: Naive implementation of autoincrement in DB

File: src/main/scala/com/krykowski/repository/TaskInMemoryRepository.scala [hunk 4]
```diff
@@ -0,0 +1,53 @@
+package com.krykowski.repository
+
+import cats.effect.IO
+import com.krykowski.model.{Canceled, Task, TaskNotFoundError}
+import org.slf4j.LoggerFactory
+
+class TaskInMemoryRepository extends TaskRepository {
+  private val logger = LoggerFactory.getLogger(getClass)
+
+  var tasks: List[Task] = List()
```
Comment (marcinkrykowski) [inline src/main/scala/com/krykowski/repository/TaskInMemoryRepository.scala:10]: Awful `var`.

File: src/main/scala/com/krykowski/repository/TaskInMemoryRepository.scala [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+  def getTasks: List[Task] = {
+    logger.info("Fetching all tasks from DB")
+    tasks
+  }
+
+  def getTask(id: Long): IO[Either[TaskNotFoundError.type, Task]] = {
+    logger.info(s"Fetching task with id: $id from DB")
+    IO(tasks.filter(task => task.id.get == id) match {
+      case task :: _ => Right(task)
```
Comment (marcinkrykowski) [inline src/main/scala/com/krykowski/repository/TaskInMemoryRepository.scala:33]: In the list there should only be one task with given ID but still this piece is not very safe.

File: src/main/scala/com/krykowski/repository/TaskInMemoryRepository.scala [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    logger.info(s"Cancelling task with id: $id in DB")
+    getTask(id).map {
+      case Left(_) => Left(TaskNotFoundError)
+      case Right(task) =>
+        val newTask = task.copy(
+          id = task.id,
+          status = Canceled,
+          jsonFileUri = task.jsonFileUri
+        )
+        tasks = newTask :: tasks.filterNot(_.id.contains(id))
```
Comment (marcinkrykowski) [inline src/main/scala/com/krykowski/repository/TaskInMemoryRepository.scala:48]: Should be extracted and refactored. What we do here is we take the task for given if from the storage, copy its values and then remove the original task from the list and finally adding the new task.

File: src/main/scala/com/krykowski/service/TaskServiceImpl.scala [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+      Task(None, Running, jsonFileName)
+      val downloadedCsv = Paths.get(s"${csv.name}.csv")
+      transformToJson(downloadedCsv, jsonFileName)
+      Task(None, Done, jsonFileName)
+    }
+    else {
+      Task(None, Scheduled, jsonFileName)
+    }
+    taskRepository.createTask(task)
+  }
```
Comment (marcinkrykowski) [inline src/main/scala/com/krykowski/service/TaskServiceImpl.scala:46]: Way too big method. Could have been refactored into smaller pieces.

File: src/main/scala/com/krykowski/service/TaskServiceImpl.scala [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+  implicit private val cs = IO.contextShift(global)
+  implicit private val cf = IO.ioConcurrentEffect
+
+  override def createTask(csv: CsvFileLocation): IO[Long] = {
+    getCsvFile(csv.uri) match {
+      case Left(ex) =>
+        logger.info(s"Couldn't download the file: ${ex.toString}")
+        Left(ex)
+      case Right(_) => fileDownloader(csv.uri, s"${csv.name}.csv")
+    }
```
Comment (marcinkrykowski) [inline src/main/scala/com/krykowski/service/TaskServiceImpl.scala:31]: What happens with this value? Nothing :( It's an Unit operation that unfortunately might fail e.g. `fileDownloader`

File: src/main/scala/com/krykowski/service/TaskServiceImpl.scala [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    val numberOfRunningTasks = taskRepository.getTasks.count(task => task.status == Running)
+    logger.info(s"Number of running tasks: $numberOfRunningTasks")
+    val jsonFileName = s"${csv.name}.json"
+    logger.info(s"Json file will be saved to: $jsonFileName")
+    val task = if (numberOfRunningTasks < 2) {
+      Task(None, Running, jsonFileName)
```
Comment (marcinkrykowski) [inline src/main/scala/com/krykowski/service/TaskServiceImpl.scala:37]: This task is not used anywhere nor returned.

File: src/main/scala/com/krykowski/service/TaskServiceImpl.scala [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+  private def transformToJson(path: Path, pathToSave: String) = {
+    val output = new File(s"$pathToSave")
+    logger.info(s"Transforming csv from: ${path.toString} to json: ${output.toString}")
+    Blocker[IO].use(blocker => {
+      fs2.io.file.readAll[IO](
+        path, blocker, 4096
+      ).through(fs2.text.utf8Decode)
+        .through(fs2.text.lines)
+        .map(_.split(","))
+        .map(_.asJson.noSpaces)
```
Comment (marcinkrykowski) [inline src/main/scala/com/krykowski/service/TaskServiceImpl.scala:76]: This method should contain an accumulator for counting the number of lines.

File: src/main/scala/com/krykowski/service/TaskServiceImpl.scala [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+      ).through(fs2.text.utf8Decode)
+        .through(fs2.text.lines)
+        .compile.toList
+    }).unsafeRunSync()
+  }
+
+  private def fileDownloader(url: String, filename: String) = {
+    logger.info(s"Downloading file from $url and saving to $filename")
+    new URL(url) #> new File(filename) !!
+  }
```
Comment (marcinkrykowski) [inline src/main/scala/com/krykowski/service/TaskServiceImpl.scala:98]: Very unsafe and even dangerous code.

File: src/test/scala/com/krykowski/api/TaskApiSpec.scala [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+      val task1 = Task(Some(id1), Scheduled, "/someUri")
+      val id2 = 2
+      val task2 = Task(Some(id2), Running, "/someOtherUri")
+      val tasks = List(task1, task2)
+      (taskService.getAllTasks _).when().returns(tasks)
+
+      val response = serve(Request[IO](GET, uri"/task"))
+      response.status shouldBe Status.Ok
+      response.as[String].unsafeRunSync() shouldBe
+        s"""[{"id":${task1.id.get},"status":"${task1.status}","jsonFileUri":"${task1.jsonFileUri}"},{"id":${task2.id.get},"status":"${task2.status}","jsonFileUri":"${task2.jsonFileUri}"}]"""
```
Comment (marcinkrykowski) [inline src/test/scala/com/krykowski/api/TaskApiSpec.scala:31]: Json formatting should be added to the api layer.

File: src/test/scala/com/krykowski/repository/TaskRepositorySpec.scala [hunk 13]
```diff
@@ -0,0 +1,7 @@
+package com.krykowski.repository
+
+import org.scalamock.scalatest.MockFactory
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpec
+
+class TaskRepositorySpec extends AnyWordSpec with MockFactory with Matchers {}
```
Comment (marcinkrykowski) [inline src/test/scala/com/krykowski/repository/TaskRepositorySpec.scala:7]: No tests.

File: src/test/scala/com/krykowski/service/TaskServiceSpec.scala [hunk 14]
```diff
@@ -0,0 +1,7 @@
+package com.krykowski.service
+
+import org.scalamock.scalatest.MockFactory
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpec
+
+class TaskServiceSpec extends AnyWordSpec with MockFactory with Matchers {}
```
Comment (marcinkrykowski) [inline src/test/scala/com/krykowski/service/TaskServiceSpec.scala:7]: No tests.

**Review by marcinkrykowski — COMMENTED** [review]
Also, it seems that I haven't implemented a task executor that would add tasks to a queue and then execute them using some scheduling policy. Some parts are missing, but as I mentioned it's not because I don't know how to implement them, but rather I was short on time. Or maybe it's just a lie I tell myself... 🤔

File: src/main/scala/com/krykowski/service/TaskService.scala [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+import com.krykowski.model.{CsvFileLocation, Task, TaskNotFoundError}
+
+trait TaskService {
+  def createTask(csv: CsvFileLocation): IO[Long]
+
+  def getAllTasks(): List[Task]
+
+  def cancelTask(taskId: Long): IO[Either[TaskNotFoundError.type, Unit]]
+
+  def getJson(taskId: Long): IO[Either[model.TaskNotFoundError.type, List[String]]]
```
Comment (marcelmojzis) [inline src/main/scala/com/krykowski/service/TaskService.scala:14]: One of critical requirements of the challenge is, that processed files might not fit into the memory and they must be streamed. As `getJson` returns `List[String]`, it loads whole file into memory. We expected a stream of byte chunks/lines here.

File: src/main/scala/com/krykowski/service/TaskServiceImpl.scala [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+        .map(_.split(","))
+        .map(_.asJson.noSpaces)
+        .intersperse("\n")
+        .through(fs2.text.utf8Encode)
+        .through(fs2.io.file.writeAll(output.toPath, blocker))
+        .compile.drain
+    }).unsafeRunSync()
+  }
+
+  private def streamJsonFile(jsonFile: Path) = {
```
Comment (marcelmojzis) [inline src/main/scala/com/krykowski/service/TaskServiceImpl.scala:84]: The method actually loads the content of the file into memory, the name is misleading. Personally I always annotate return type also for private functions. In this case it is `String`, not `Stream[IO, Byte]` which I would expect for a function with name `streamJsonFile`.

File: src/main/scala/com/krykowski/service/TaskServiceImpl.scala [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+        jsonFile, blocker, 4096
+      ).through(fs2.text.utf8Decode)
+        .through(fs2.text.lines)
+        .compile.toList
+    }).unsafeRunSync()
+  }
+
+  private def fileDownloader(url: String, filename: String) = {
+    logger.info(s"Downloading file from $url and saving to $filename")
+    new URL(url) #> new File(filename) !!
```
Comment (marcelmojzis) [inline src/main/scala/com/krykowski/service/TaskServiceImpl.scala:97]: Sync + starts a new process. We expected exact opposite...

File: src/main/scala/com/krykowski/service/TaskServiceImpl.scala [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+import java.net.URL
+import java.nio.file.{Path, Paths}
+
+class TaskServiceImpl(taskRepository: TaskRepository) extends TaskService {
+  private val logger = LoggerFactory.getLogger(getClass)
+
+  implicit private val cs = IO.contextShift(global)
+  implicit private val cf = IO.ioConcurrentEffect
+
+  override def createTask(csv: CsvFileLocation): IO[Long] = {
```
Comment (marcelmojzis) [inline src/main/scala/com/krykowski/service/TaskServiceImpl.scala:25]: A task is an operation which includes downloading a CSV file, converting it to JSON and storing it in filesystem. Up to 2 tasks can run concurrently.

File: src/main/scala/com/krykowski/service/TaskServiceImpl.scala [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+    getCsvFile(csv.uri) match {
```
Comment (marcelmojzis) [inline src/main/scala/com/krykowski/service/TaskServiceImpl.scala:26]: File is downloaded without checking if other tasks are running. Calling thread is blocked. The is no way how to cancel a running task at this stage.

File: src/main/scala/com/krykowski/service/TaskServiceImpl.scala [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+      case Left(ex) =>
+        logger.info(s"Couldn't download the file: ${ex.toString}")
+        Left(ex)
+      case Right(_) => fileDownloader(csv.uri, s"${csv.name}.csv")
```
Comment (marcelmojzis) [inline src/main/scala/com/krykowski/service/TaskServiceImpl.scala:30]: File is downloaded for the second time...

File: src/main/scala/com/krykowski/service/TaskServiceImpl.scala [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+    }
+    val numberOfRunningTasks = taskRepository.getTasks.count(task => task.status == Running)
```
Comment (marcelmojzis) [inline src/main/scala/com/krykowski/service/TaskServiceImpl.scala:32]: `numberOfRunningTasks` doesn't represent number of running tasks, it represents the number of tasks that reached line 45 of this file and at that point file is already downloaded and converted.

File: src/main/scala/com/krykowski/service/TaskServiceImpl.scala [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+    logger.info(s"Number of running tasks: $numberOfRunningTasks")
+    val jsonFileName = s"${csv.name}.json"
+    logger.info(s"Json file will be saved to: $jsonFileName")
+    val task = if (numberOfRunningTasks < 2) {
+      Task(None, Running, jsonFileName)
+      val downloadedCsv = Paths.get(s"${csv.name}.csv")
+      transformToJson(downloadedCsv, jsonFileName)
```
Comment (marcelmojzis) [inline src/main/scala/com/krykowski/service/TaskServiceImpl.scala:39]: Blocking, no way how to cancel task

File: src/main/scala/com/krykowski/service/TaskServiceImpl.scala (hunk 18 above)
Comment (marcelmojzis) [inline src/main/scala/com/krykowski/service/TaskServiceImpl.scala:25]: All effects that matter here live outside IO.

**Pull request closed without merging**